### ISSUE TITLE
openssl DH deprecation warnings

### ISSUE THREAD

**Issue by ckerr** [transmission/transmission#3004]
I'm seeing these after upgrading to openssl 3.0.2-0ubuntu1

```
[9/183] Building CXX object libtransmission/CMakeFiles/transmission.dir/crypto-utils-openssl.cc.o
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:259:18: warning: 'DH_new' is deprecated [clang-diagnostic-deprecated-declarations]
    DH* handle = DH_new();
                 ^
/usr/include/openssl/dh.h:199:1: note: 'DH_new' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 DH *DH_new(void);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:264:51: warning: 'DH_set0_pqg' is deprecated [clang-diagnostic-deprecated-declarations]
    if (!check_pointer(p) || !check_pointer(g) || DH_set0_pqg(handle, p, nullptr, g) == 0)
                                                  ^
/usr/include/openssl/dh.h:255:1: note: 'DH_set0_pqg' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:268:9: warning: 'DH_free' is deprecated [clang-diagnostic-deprecated-declarations]
        DH_free(handle);
        ^
/usr/include/openssl/dh.h:200:1: note: 'DH_free' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 void DH_free(DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:284:5: warning: 'DH_free' is deprecated [clang-diagnostic-deprecated-declarations]
    DH_free(handle);
    ^
/usr/include/openssl/dh.h:200:1: note: 'DH_free' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 void DH_free(DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:294:5: warning: 'DH_set_length' is deprecated [clang-diagnostic-deprecated-declarations]
    DH_set_length(handle, private_key_length * 8);
    ^
/usr/include/openssl/dh.h:269:1: note: 'DH_set_length' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_set_length(DH *dh, long length);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:296:23: warning: 'DH_generate_key' is deprecated [clang-diagnostic-deprecated-declarations]
    if (!check_result(DH_generate_key(handle)))
                      ^
/usr/include/openssl/dh.h:223:1: note: 'DH_generate_key' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_generate_key(DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:302:5: warning: 'DH_get0_key' is deprecated [clang-diagnostic-deprecated-declarations]
    DH_get0_key(handle, &my_public_key, nullptr);
    ^
/usr/include/openssl/dh.h:256:1: note: 'DH_get0_key' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 void DH_get0_key(const DH *dh, const BIGNUM **pub_key,
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:305:25: warning: 'DH_size' is deprecated [clang-diagnostic-deprecated-declarations]
    int const dh_size = DH_size(handle);
                        ^
/usr/include/openssl/dh.h:203:1: note: 'DH_size' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_size(const DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:330:25: warning: 'DH_size' is deprecated [clang-diagnostic-deprecated-declarations]
    int const dh_size = DH_size(handle);
                        ^
/usr/include/openssl/dh.h:203:1: note: 'DH_size' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_size(const DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:332:35: warning: 'DH_compute_key' is deprecated [clang-diagnostic-deprecated-declarations]
    int const secret_key_length = DH_compute_key(ret->key, other_key, handle);
                                  ^
/usr/include/openssl/dh.h:224:1: note: 'DH_compute_key' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_compute_key(unsigned char *key,
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:259:18: warning: 'DH_new' is deprecated [-Wdeprecated-declarations]
    DH* handle = DH_new();
                 ^
/usr/include/openssl/dh.h:199:1: note: 'DH_new' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 DH *DH_new(void);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:264:51: warning: 'DH_set0_pqg' is deprecated [-Wdeprecated-declarations]
    if (!check_pointer(p) || !check_pointer(g) || DH_set0_pqg(handle, p, nullptr, g) == 0)
                                                  ^
/usr/include/openssl/dh.h:255:1: note: 'DH_set0_pqg' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:268:9: warning: 'DH_free' is deprecated [-Wdeprecated-declarations]
        DH_free(handle);
        ^
/usr/include/openssl/dh.h:200:1: note: 'DH_free' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 void DH_free(DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:284:5: warning: 'DH_free' is deprecated [-Wdeprecated-declarations]
    DH_free(handle);
    ^
/usr/include/openssl/dh.h:200:1: note: 'DH_free' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 void DH_free(DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:294:5: warning: 'DH_set_length' is deprecated [-Wdeprecated-declarations]
    DH_set_length(handle, private_key_length * 8);
    ^
/usr/include/openssl/dh.h:269:1: note: 'DH_set_length' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_set_length(DH *dh, long length);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:296:23: warning: 'DH_generate_key' is deprecated [-Wdeprecated-declarations]
    if (!check_result(DH_generate_key(handle)))
                      ^
/usr/include/openssl/dh.h:223:1: note: 'DH_generate_key' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_generate_key(DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:302:5: warning: 'DH_get0_key' is deprecated [-Wdeprecated-declarations]
    DH_get0_key(handle, &my_public_key, nullptr);
    ^
/usr/include/openssl/dh.h:256:1: note: 'DH_get0_key' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 void DH_get0_key(const DH *dh, const BIGNUM **pub_key,
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:305:25: warning: 'DH_size' is deprecated [-Wdeprecated-declarations]
    int const dh_size = DH_size(handle);
                        ^
/usr/include/openssl/dh.h:203:1: note: 'DH_size' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_size(const DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:330:25: warning: 'DH_size' is deprecated [-Wdeprecated-declarations]
    int const dh_size = DH_size(handle);
                        ^
/usr/include/openssl/dh.h:203:1: note: 'DH_size' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_size(const DH *dh);
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
/home/charles/src/transmission/libtransmission/crypto-utils-openssl.cc:332:35: warning: 'DH_compute_key' is deprecated [-Wdeprecated-declarations]
    int const secret_key_length = DH_compute_key(ret->key, other_key, handle);
                                  ^
/usr/include/openssl/dh.h:224:1: note: 'DH_compute_key' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 int DH_compute_key(unsigned char *key,
^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
10 warnings generated.
```

**Comment by ckerr** [comment]
https://stackoverflow.com/questions/71551116/openssl-3-diffie-hellman-key-exchange-c

**Comment by ckerr** [comment]
Fixed by 2bcab6be7e39920247dbf874a58c6d6a2f2d59fe
